### PR TITLE
[rush] Support for PNPM 4 (Experimental)

### PR DESCRIPTION
**Pull request by sachinjoseph**
Tested rush install and rush build on web-build-tools.

Reference: https://github.com/pnpm/pnpm/releases/tag/v4.0.0

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -9,6 +9,9 @@ yarn-error.log*
 *.seed
 *.pid.lock
 
+# Visual Studio Code
+.vscode
```
Comment (octogonz) [inline .gitignore]: Should we exclude these folders entirely?  Currently we have added some **.vscode/launch.json** files to Git as templates.

File: apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts [hunk 2]
```diff
@@ -32,5 +36,17 @@ export class PnpmPackageManager extends PackageManager {
...
+    if (parsedVersion.major <= 2) {
+      // node_modules/.shrinkwrap.yaml
+      this.internalShrinkwrapRelativePath = path.join('node_modules', '.shrinkwrap.yaml');
+    } else if (parsedVersion.major <= 3) {
+      // node_modules/.pnpm-lock.yaml
+      this.internalShrinkwrapRelativePath = path.join('node_modules', '.pnpm-lock.yaml');
+    } else {
+      // node_modules/.pnpm/lock.yaml
+      // See https://github.com/pnpm/pnpm/releases/tag/v4.0.0 for more details.
+      this.internalShrinkwrapRelativePath = path.join('node_modules', '.pnpm', 'lock.yaml');
```
Comment (octogonz) [inline apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts]: 🤨 Did the also rename the main shrinkwrap file?  Or is this just for the copy under the node_modules folder?

File: common/changes/@microsoft/rush/sachinjoseph-SupportForPnpm4_2019-10-09-02-14.json [hunk 3]
```diff
@@ -0,0 +1,11 @@
+{
+  "changes": [
+    {
+      "packageName": "@microsoft/rush",
+      "comment": "Support PNPM 4 on Rush",
```
Comment (octogonz) [inline common/changes/@microsoft/rush/sachinjoseph-SupportForPnpm4_2019-10-09-02-14.json]: "Add support for PNPM 4.x"

How confident are we that it works reliably?  If this is not very well tested, maybe say "Add support for PNPM 4.x (Experimental)"

File: .gitignore (hunk 1 above)
Comment (sachinjoseph) [reply .gitignore]: Even though I don't feel very strongly about this change, launch.json shows up as modified every time someone changes some local debug config - which is quite an inconvenience. How about we explicitly mark template folders as include in .gitignore?

File: apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts (hunk 2 above)
Comment (apostolisms) [reply apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts]: just the copy under node_modules

File: common/changes/@microsoft/rush/sachinjoseph-SupportForPnpm4_2019-10-09-02-14.json (hunk 3 above)
Comment (sachinjoseph) [reply common/changes/@microsoft/rush/sachinjoseph-SupportForPnpm4_2019-10-09-02-14.json]: Yes, let's add the experimental tag for now. It's possible I am missing something.

File: apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts (hunk 2 above)
Comment (sachinjoseph) [reply apps/rush-lib/src/api/packageManager/PnpmPackageManager.ts]: Only the one in node_modules folder.

File: apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts [hunk 4]
```diff
@@ -207,13 +212,8 @@ export class PnpmLinkManager extends BaseLinkManager {
 
     // tslint:disable-next-line:max-line-length
     // e.g.: C:\wbt\common\temp\node_modules\.local\C%3A%2Fwbt%2Fcommon%2Ftemp%2Fprojects%2Fapi-documenter.tgz\node_modules
```
Comment (D4N14L) [inline apps/rush-lib/src/logic/pnpm/PnpmLinkManager.ts]: This doesn't necessarily match anymore. Maybe make a note on the comment?

File: .gitignore (hunk 1 above)
Comment (iclanton) [reply .gitignore]: I'm in favor of ignoring.

**Pull request merged**